metadata-translation/notebooks/repl_validation_referential_integrity-1715162638.ipynb

```diff
@@ -0,0 +1,686 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b962e3c8-a346-49c5-8470-915f3cf9eb07",
+   "metadata": {},
+   "source": [
+    "Materialize `alldocs` collection, associating all inherited classes with document via `type` field.\n",
+    "\n",
+    "> **TODO:** Clarify the above sentence."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "b2e618f3-78b9-42b6-8ea9-63d080b1b0f6",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [
+    {
+     "data": {
+      "application/vnd.jupyter.widget-view+json": {
+       "model_id": "e69c4fd820114e33b11ebae47f9f3e4d",
+       "version_major": 2,
+       "version_minor": 0
+      },
+      "text/plain": [
+       "  0%|          | 0/224995 [00:00<?, ?it/s]"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "refreshed `alldocs` collection\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Drop any existing `alldocs` collection (e.g. from previous use of this notebook).\n",
+    "mdb.alldocs.drop()\n",
+    "\n",
+    "# Set up progress bar\n",
+    "n_docs_total = sum(mdb[name].estimated_document_count() for name in collection_names)\n",
+    "pbar = tqdm(total=n_docs_total)\n",
+    "\n",
+    "# for each collection name\n",
+    "for coll_name in collection_names:\n",
+    "    pbar.set_description(f\"processing {coll_name}...\")\n",
+    "    # for each doc in collection, remove the mongo-generated '_id' field\n",
+    "    try:\n",
+    "        nmdcdb = NMDCDatabase(**{coll_name: [dissoc(mdb[coll_name].find_one(), '_id')]})\n",
+    "    except ValueError as e:\n",
+    "        print(f\"no {coll_name}!\")\n",
+    "        raise e\n",
+    "\n",
+    "    # Calculate class_hierarchy_as_list once per collection.\n",
+    "    #\n",
+    "    # Note: This seems to assume that the class hierarchy is identical for each document\n",
+    "    #       in a given collection, which may not be the case since a collection whose\n",
+    "    #       range is a \"parent\" class can store instances of descendant classes (and the\n",
+    "    #       class hierarchy of the latter would differ from that of the former).\n",
+    "    #\n",
+    "    exemplar = getattr(nmdcdb, coll_name)[0]  # get first instance (i.e. document) in list\n",
+    "    newdoc_type: list[str] = class_hierarchy_as_list(exemplar)\n",
+    "    \n",
+    "    # For each document in this collection, replace the value of the `type` field with\n",
+    "    # a _list_ of the document's own class and ancestor classes, remove the `_id` field,\n",
+    "    # and insert the resulting document into the `alldocs` collection. Note that we are not\n",
+    "    # relying on the original value of the `type` field, since it's unreliable (see below).\n",
+    "    \n",
+    "    # NOTE: `type` is currently a string, does not exist for all classes, and can have typos. \n",
```

Comment:
`type` currently being a string caused a lot of problems with re-iding. Current recommendation is infer type from collection name until we are on Berkeley schema.